enhancements/authentication/synchronizing-user-groups-on-login.md

```diff
@@ -0,0 +1,238 @@
+   the group appears in the configured claims
+3. on login, user gets added to an existing group that appears in the
+   retrieved group claims
+4. on login, user gets removed from any existing group that was created
+   for the given identity provider (per annotation) when the group
+   no longer appears among those retrieved from the OIDC
+
+
+### Follow-up Work
+
+To synchronize groups in a time-based manner, an additional controller in the
+oauth-server would be necessary. OIDC does not provide a standardized group
+membership and reverse group membership lookups. That means that such a controller
+would have to cache the access tokens (and possibly refresh tokens) retrieved
+from the OIDC after a successful user login, and use them to query the UserInfo
+endpoint. This could be troublesome, however, because:
+
+1. the access token we get from the OIDC might be shortlived, if there are no
+   refresh tokens we can't synchronize the groups for the given user for too long
+2. we might end up generating a lot of traffic to the OIDC provider
+3. the OIDC may not expose/implement the UserInfo endpoint
+
+### Graduation Criteria
+
+The feature should be considered mature once it lands.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Drawbacks
+
+There are two drawbacks to the current solution:
+
+1. the groups go out of sync between logins
+  - This is quite a common problem with all systems that receive delegated identities.
+    Third party vendors don't usually provide a standardized watch channel that could be
+    effectively and reliably observed, and so most systems either only synchronize on user
+    login or attempt to perform group retrievals in regular intervals, but real real-time
+    synchronization is usually impossible.
+2. the groups are assigned to a user, not an `<identity provider:user>` touple
+  - Some configurations allow a user to be sourced by multiple identity providers, where
+    each of the providers might have different groups configured. The implemenentation
+    described in this enhancement does not distinguish the groups when the user attempts
+    to use their token and just uses all the OpenShift groups the user appears in.
+
+## Alternatives
+
+The OAuthAccessToken, User and Groups are the three things that the current OpenShift
+authenticator retrieves during token validation.
+
+To solve the issue of the user only appearing in the groups of
+`union(current identity provider groups, OpenShift groups)`, we could take an
+alternative approach and only synchronize the groups into the OAuthAccessToken for
+the given user and session, since, as noted above, the access token is one of the
+three objects introspected during authentication, and is the only one of the three
+that is sure to describe the current session.
+
+Having the groups stored within the access token would likely make group
```

Comment:
Also, the Access Token could explode in its size if a user belongs to gazillion groups at the same time (which is very likely).

Reply:
True